src/queue/descriptor_chain.rs

```diff
@@ -0,0 +1,1022 @@
+            if rem < vs.len() {
+                true
+            } else {
+                rem -= vs.len();
+                false
+            }
+        });
+
+        if let Some(at) = pos {
+            let mut other = self.buffers.split_off(at);
+
+            if rem > 0 {
+                // There must be at least one element in `other` because we checked
+                // its `size` value in the call to `position` above.
+                let front = other.pop_front().expect("empty VecDeque after split");
+                let (start, end) = front.split_at(rem).map_err(Error::VolatileMemoryError)?;
+                self.buffers.push_back(start);
+                other.push_front(end);
+            }
+
+            Ok(DescriptorChainConsumer {
+                buffers: other,
+                bytes_consumed: 0,
+            })
+        } else if rem == 0 {
+            Ok(DescriptorChainConsumer {
+                buffers: VecDeque::new(),
+                bytes_consumed: 0,
+            })
+        } else {
+            Err(Error::DescriptorChainSplitOOB(offset))
+        }
+    }
+}
+
+fn collect_desc_chain_buffers<M: GuestAddressSpace>(
+    mem: &GuestMemoryMmap,
+    desc_chain_iter: DescriptorChainRwIter<M>,
+) -> Result<VecDeque<VolatileSlice<'_>>> {
+    let mut total_len: usize = 0;
+    let mut buffers = VecDeque::new();
+    for desc in desc_chain_iter {
+        // Verify that summing the descriptor sizes does not overflow.
+        // This can happen if a driver tricks a device into reading more data than
+        // fits in a `usize`.
+        total_len = total_len
+            .checked_add(desc.len() as usize)
+            .ok_or(Error::DescriptorChainOverflow)?;
+
+        let mut len = desc.len() as usize;
+        let mut addr = desc.addr();
+        while len > 0 {
+            let region = mem.find_region(addr).ok_or(Error::FindMemoryRegion)?;
+            let offset = addr
+                .checked_sub(region.start_addr().raw_value())
+                .unwrap()
+                .raw_value() as usize;
+            let buf_len = cmp::min(region.deref().len() - offset, len);
+            let buf = region
+                .deref()
+                .get_slice(offset, buf_len)
+                .map_err(Error::VolatileMemoryError)?;
+            buffers.push_back(buf);
+            len -= buf_len;
+            addr = addr
+                .checked_add(buf_len as _)
+                .ok_or(Error::DescriptorChainOverflow)?;
+        }
+    }
+
+    Ok(buffers)
+}
+
+/// Provides high-level interface over the sequence of memory regions
+/// defined by readable descriptors in the descriptor chain.
+///
+/// Note that virtio spec requires driver to place any device-writable
+/// descriptors after any device-readable descriptors (2.6.4.2 in Virtio Spec v1.1).
+/// Reader will skip iterating over descriptor chain when first writable
+/// descriptor is encountered.
+#[derive(Clone, Debug)]
+pub struct Reader<'a> {
```

Comment:
Is there any way to get address and length of underlying descriptor through reader/writer? It would be great to have addr and len methods.

Reply:
I guess we could add a Reader::get_location() to return the current descriptor (addr, remaining_len). Is that what you would like? Perhaps you can work a patch on top? thanks

Reply:
(imho, this kind of advanced function is out of scope of Reader/Writer helper though)

Reply:
The underlying address and length are needed for virtio blk and scsi as `DescriptorChain` in this crate does not have `addr` and `len` fields but the cloud-hypervisor's vm-virtio's  `DescriptorChain` does. Maybe this should be a separate issue.

Reply:
Yes, I think it make sense to make it a separate issue.